helix-rest/src/main/java/org/apache/helix/rest/metadatastore/accessor/ZkRoutingDataWriter.java

```diff
@@ -0,0 +1,256 @@
+              _namespace, realm, shardingKey);
+          return false;
+        }
+      }
+
+      // Add the sharding key to an empty ZNRecord
+      ZNRecord znRecord;
+      try {
+        znRecord =
+            _zkClient.readData(MetadataStoreRoutingConstants.ROUTING_DATA_PATH + "/" + realm);
+      } catch (Exception e) {
+        LOG.error(
+            "Failed to read the realm ZNRecord in addShardingKey()! Namespace: {}, Realm: {}, ShardingKey: {}",
+            _namespace, realm, shardingKey, e);
+        return false;
+      }
+      znRecord.setListField(MetadataStoreRoutingConstants.ZNRECORD_LIST_FIELD_KEY,
+          Collections.singletonList(shardingKey));
+      try {
+        _zkClient
+            .writeData(MetadataStoreRoutingConstants.ROUTING_DATA_PATH + "/" + realm, znRecord);
+      } catch (Exception e) {
+        LOG.error(
+            "Failed to write the realm ZNRecord in addShardingKey()! Namespace: {}, Realm: {}, ShardingKey: {}",
+            _namespace, realm, shardingKey, e);
+        return false;
+      }
+      return true;
+    }
+
+    // TODO: Forward the request to leader
+    return true;
+  }
+
+  @Override
+  public synchronized boolean deleteShardingKey(String realm, String shardingKey) {
+    if (_leaderElection.isLeader()) {
+      if (_zkClient.isClosed()) {
+        throw new IllegalStateException("ZkClient is closed!");
+      }
+      ZNRecord znRecord =
+          _zkClient.readData(MetadataStoreRoutingConstants.ROUTING_DATA_PATH + "/" + realm, true);
+      if (znRecord == null || !znRecord
+          .getListField(MetadataStoreRoutingConstants.ZNRECORD_LIST_FIELD_KEY)
+          .contains(shardingKey)) {
+        // This realm does not exist or shardingKey doesn't exist. Return true!
+        return true;
+      }
+      znRecord.getListField(MetadataStoreRoutingConstants.ZNRECORD_LIST_FIELD_KEY)
+          .remove(shardingKey);
+      // Overwrite this ZNRecord with the sharding key removed
+      try {
+        _zkClient
+            .writeData(MetadataStoreRoutingConstants.ROUTING_DATA_PATH + "/" + realm, znRecord);
+      } catch (Exception e) {
+        LOG.error(
+            "Failed to write the data back in deleteShardingKey()! Namespace: {}, Realm: {}, ShardingKey: {}",
+            _namespace, realm, shardingKey, e);
+        return false;
+      }
+      return true;
+    }
+
+    // TODO: Forward the request to leader
+    return true;
+  }
+
+  @Override
+  public synchronized boolean setRoutingData(Map<String, List<String>> routingData) {
+    if (_leaderElection.isLeader()) {
+      if (_zkClient.isClosed()) {
+        throw new IllegalStateException("ZkClient is closed!");
+      }
+      if (routingData == null) {
+        throw new IllegalArgumentException("routingData given is null!");
+      }
+
+      // Remove existing routing data
+      for (String zkRealm : _zkClient
+          .getChildren(MetadataStoreRoutingConstants.ROUTING_DATA_PATH)) {
```

Comment:
Nit: can we recursively delete `MetadataStoreRoutingConstants.ROUTING_DATA_PATH` here?

Reply:
Deletion of MetadataStoreRoutingConstants.ROUTING_DATA_PATH is not necessary, and it could cause ZkNoNodeException.